File: app/src/main/java/pl/karol202/paintplus/tool/shape/star/StarProperties.java
/*
 * Copyright 2017 karol-202
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package pl.karol202.paintplus.tool.shape.star;

import android.os.Bundle;
import android.support.design.widget.TextInputLayout;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.*;
import pl.karol202.paintplus.R;
import pl.karol202.paintplus.tool.shape.Join;
import pl.karol202.paintplus.tool.shape.JoinAdapter;
import pl.karol202.paintplus.tool.shape.ShapeProperties;
import pl.karol202.paintplus.util.SeekBarTouchListener;

public class StarProperties extends ShapeProperties
{
	private final int MIN_CORNERS = 3;
	private final int MAX_CORNERS = 20;
	
	private ShapeStar star;
	private String errorToFew;
	private String errorToMany;
	private JoinAdapter adapter;
	
	private View view;
	private ImageButton buttonMinusCorners;
	private ImageButton buttonPlusCorners;
	private TextInputLayout editLayoutCorners;
	private EditText editCorners;
	private CheckBox checkFill;
	private SeekBar seekBarWidth;
	private TextView textWidth;
	private Spinner spinnerJoin;
	
	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
	{
		super.onCreateView(inflater, container, savedInstanceState);
		view = inflater.inflate(R.layout.properties_star, container, false);
		
		star = (ShapeStar) shape;
		errorToFew = getActivity().getString(R.string.error_polygon_too_few_sides);
		errorToMany = getActivity().getString(R.string.error_polygon_too_many_sides);
		adapter = new JoinAdapter(getActivity());
		
		buttonMinusCorners = view.findViewById(R.id.button_minus_star_corners);
		buttonMinusCorners.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v)
			{
				if(getSides() > MIN_CORNERS) editCorners.setText(String.valueOf(getSides() - 1));
			}
		});
		
		buttonPlusCorners = view.findViewById(R.id.button_plus_star_corners);
		buttonPlusCorners.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v)
			{
				if(getSides() < MAX_CORNERS) editCorners.setText(String.valueOf(getSides() + 1));
			}
		});
		
		editLayoutCorners = view.findViewById(R.id.edit_layout_star_corners);
		editLayoutCorners.setHintEnabled(false);
		
		editCorners = editLayoutCorners.getEditText();
		if(editCorners == null) throw new RuntimeException("TextInputLayout must contain EditText.");
		editCorners.setText(String.valueOf(star.getCorners()));
		editCorners.addTextChangedListener(new TextWatcher() {
			@Override
			public void beforeTextChanged(CharSequence s, int start, int count, int after) { }
			
			@Override
			public void onTextChanged(CharSequence s, int start, int before, int count) { }
			
			@Override
			public void afterTextChanged(Editable s)
			{
				int sides = getSides();
				if(sides < MIN_CORNERS) editLayoutCorners.setError(errorToFew);
				else if(sides > MAX_CORNERS) editLayoutCorners.setError(errorToMany);
				else
				{
					editLayoutCorners.setErrorEnabled(false);
					star.setCorners(getSides());
				}
			}
		});
		
		checkFill = view.findViewById(R.id.check_star_fill);
		checkFill.setChecked(star.isFill());
		checkFill.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
			@Override
			public void onCheckedChanged(CompoundButton buttonView, boolean isChecked)
			{
				star.setFill(isChecked);
			}
		});
		
		seekBarWidth = view.findViewById(R.id.seek_star_width);
		seekBarWidth.setProgress(star.getLineWidth() - 1);
		seekBarWidth.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
			@Override
			public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser)
			{
				star.setLineWidth(progress + 1);
				textWidth.setText(String.valueOf(progress + 1));
			}
			
			@Override
			public void onStartTrackingTouch(SeekBar seekBar) { }
			
			@Override
			public void onStopTrackingTouch(SeekBar seekBar) { }
		});
		seekBarWidth.setOnTouchListener(new SeekBarTouchListener());
		
		textWidth = view.findViewById(R.id.star_width);
		textWidth.setText(String.valueOf(star.getLineWidth()));
		
		spinnerJoin = view.findViewById(R.id.spinner_star_join);
		spinnerJoin.setAdapter(adapter);
		spinnerJoin.setSelection(star.getJoin().ordinal());
		spinnerJoin.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
			@Override
			public void onItemSelected(AdapterView<?> parent, View view, int position, long id)
			{
				Join join = adapter.getItem(position);
				star.setJoin(join);
			}
			
			@Override
			public void onNothingSelected(AdapterView<?> parent) { }
		});
		
		return view;
	}
	
	private int getSides()
	{
		if(editCorners.getText().length() == 0) return 0;
		return Integer.parseInt(editCorners.getText().toString());
	}
}

File: app/src/main/java/pl/karol202/paintplus/tool/shape/star/StarProperties.kt
/*
 * Copyright 2017 karol-202
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package pl.karol202.paintplus.tool.shape.star

import android.annotation.SuppressLint
import pl.karol202.paintplus.tool.shape.JoinAdapter
import android.os.Bundle
import pl.karol202.paintplus.R
import android.view.View
import pl.karol202.paintplus.util.SeekBarTouchListener
import androidx.core.widget.addTextChangedListener
import androidx.fragment.app.Fragment
import org.koin.android.ext.android.inject
import pl.karol202.paintplus.databinding.PropertiesStarBinding
import pl.karol202.paintplus.tool.shape.Join
import pl.karol202.paintplus.util.setOnItemSelectedListener
import pl.karol202.paintplus.util.setOnValueChangeListener
import pl.karol202.paintplus.util.viewBinding

private const val MIN_CORNERS = 3
private const val MAX_CORNERS = 20

class StarProperties : Fragment(R.layout.properties_star)
{
	private val shapeStar by inject<ShapeStar>()

	private val views by viewBinding(PropertiesStarBinding::bind)

	private var errorToFew = getString(R.string.error_polygon_too_few_sides)
	private var errorToMany = getString(R.string.error_polygon_too_many_sides)

	private val corners get() = views.editTextStarCorners.text?.toString()?.toIntOrNull() ?: 0

	@SuppressLint("ClickableViewAccessibility")
	override fun onViewCreated(view: View, savedInstanceState: Bundle?)
	{
		views.buttonMinusStarCorners.setOnClickListener { onCornersDecremented() }

		views.buttonPlusStarCorners.setOnClickListener { onCornersIncremented() }

		views.editLayoutStarCorners.isHintEnabled = false

		views.editTextStarCorners.setText(shapeStar.corners.toString())
		views.editTextStarCorners.addTextChangedListener { onCornersChanged() }

		views.checkStarFill.isChecked = shapeStar.isFilled
		views.checkStarFill.setOnCheckedChangeListener { _, checked: Boolean -> onFillChanged(checked) }

		views.seekStarWidth.progress = shapeStar.lineWidth - 1
		views.seekStarWidth.setOnValueChangeListener { progress -> onWidthChanged(progress) }
		views.seekStarWidth.setOnTouchListener(SeekBarTouchListener())

		views.starWidth.text = shapeStar.lineWidth.toString()

		views.spinnerStarJoin.adapter = JoinAdapter(requireContext())
		views.spinnerStarJoin.setSelection(shapeStar.join.ordinal)
		views.spinnerStarJoin.setOnItemSelectedListener { index -> onJoinChanged(index) }
	}

	private fun onCornersDecremented()
	{
		if(corners > MIN_CORNERS) views.editTextStarCorners.setText((corners - 1).toString())
	}

	private fun onCornersIncremented()
	{
		if(corners < MAX_CORNERS) views.editTextStarCorners.setText((corners + 1).toString())
	}

	private fun onCornersChanged() = when
	{
		corners < MIN_CORNERS -> views.editLayoutStarCorners.error = errorToFew
		corners > MAX_CORNERS -> views.editLayoutStarCorners.error = errorToMany
		else ->
		{
			views.editLayoutStarCorners.isErrorEnabled = false
			shapeStar.corners = corners
		}
	}

	private fun onFillChanged(fill: Boolean)
	{
		shapeStar.isFilled = fill
	}

	private fun onWidthChanged(progress: Int)
	{
		shapeStar.lineWidth = progress + 1
		views.starWidth.text = (progress + 1).toString()
	}

	private fun onJoinChanged(index: Int)
	{
		shapeStar.join = Join.values()[index]
	}
}
